src/theory/strings/infer_proof_cons.cpp

```diff
@@ -0,0 +1,1043 @@
+            ps.d_rule = PfRule::RE_UNFOLD_NEG_CONCAT_FIXED;
+          }
+        }
+      }
+    }
+    break;
+    // ========================== Reduction
+    case Inference::CTN_POS:
+    case Inference::CTN_NEG_EQUAL:
+    {
+      if (ps.d_children.size() != 1)
+      {
+        break;
+      }
+      bool polarity = ps.d_children[0].getKind() != NOT;
+      Node atom = polarity ? ps.d_children[0] : ps.d_children[0][0];
+      std::vector<Node> args;
+      args.push_back(atom);
+      Node res = psb.tryStep(PfRule::STRING_EAGER_REDUCTION, {}, args);
+      if (res.isNull())
+      {
+        break;
+      }
+      // ite( contains(x,t), x = k1 ++ t ++ k2, x != t )
+      std::vector<Node> tiChildren;
+      tiChildren.push_back(ps.d_children[0]);
+      Node ctnt = psb.tryStep(
+          polarity ? PfRule::TRUE_INTRO : PfRule::FALSE_INTRO, tiChildren, {});
+      if (ctnt.isNull() || ctnt.getKind() != EQUAL)
+      {
+        break;
+      }
+      std::vector<Node> tchildren;
+      tchildren.push_back(ctnt);
+      // apply substitution { contains(x,t) -> true|false } and rewrite to get
+      // conclusion x = k1 ++ t ++ k2 or x != t.
+      if (psb.applyPredTransform(res, conc, tchildren))
+      {
+        useBuffer = true;
+      }
+    }
+    break;
+
+    case Inference::REDUCTION:
+    {
+      size_t nchild = conc.getNumChildren();
+      Node mainEq;
+      if (conc.getKind() == EQUAL)
+      {
+        mainEq = conc;
+      }
+      else if (conc.getKind() == AND && conc[nchild - 1].getKind() == EQUAL)
+      {
+        mainEq = conc[nchild - 1];
+      }
+      if (mainEq.isNull())
+      {
+        Trace("strings-ipc-red") << "Bad Reduction: " << conc << std::endl;
+        Assert(false) << "Unexpected reduction " << conc;
+        break;
+      }
+      std::vector<Node> argsRed;
+      // the left hand side of the last conjunct is the term we are reducing
+      argsRed.push_back(mainEq[0]);
+      Node red = psb.tryStep(PfRule::STRING_REDUCTION, {}, argsRed);
+      Trace("strings-ipc-red") << "Reduction : " << red << std::endl;
+      if (!red.isNull())
+      {
+        // either equal or rewrites to it
+        std::vector<Node> cexp;
+        if (psb.applyPredTransform(red, conc, cexp))
+        {
+          Trace("strings-ipc-red") << "...success!" << std::endl;
+          useBuffer = true;
+        }
+        else
+        {
+          Trace("strings-ipc-red") << "...failed to reduce" << std::endl;
+        }
+      }
+    }
+    break;
+    // ========================== code injectivity
+    case Inference::CODE_INJ:
+    {
+      ps.d_rule = PfRule::STRING_CODE_INJ;
+      Assert(conc.getKind() == OR && conc.getNumChildren() == 3
+             && conc[2].getKind() == EQUAL);
+      ps.d_args.push_back(conc[2][0]);
+      ps.d_args.push_back(conc[2][1]);
+    }
+    break;
+    // ========================== unit injectivity
+    case Inference::UNIT_INJ: { ps.d_rule = PfRule::STRING_SEQ_UNIT_INJ;
+    }
```

Comment:
```suggestion
    case Inference::UNIT_INJ: { ps.d_rule = PfRule::STRING_SEQ_UNIT_INJ;
    }
```

Formatting seems off here.

Reply:
Strangely, this is clang-formatted.